clang-tools-extra/test/clang-tidy/checkers/performance/unnecessary-value-param-templates.cpp

```diff
@@ -0,0 +1,93 @@
+// RUN: %check_clang_tidy  -std=c++14-or-later %s performance-unnecessary-value-param %t
+
+struct ExpensiveToCopyType {
+  virtual ~ExpensiveToCopyType();
+};
+
+template <typename T> void templateWithNonTemplatizedParameter(const ExpensiveToCopyType S, T V) {
+  // CHECK-MESSAGES: [[@LINE-1]]:90: warning: the const qualified parameter 'S'
+  // CHECK-MESSAGES: [[@LINE-2]]:95: warning: the parameter 'V'
+  // CHECK-FIXES: template <typename T> void templateWithNonTemplatizedParameter(const ExpensiveToCopyType& S, const T& V) {
+}
+
+void instantiatedWithExpensiveValue() {
+  templateWithNonTemplatizedParameter(
+      ExpensiveToCopyType(), ExpensiveToCopyType());
+  templateWithNonTemplatizedParameter(ExpensiveToCopyType(), 5);
+}
+
+template <typename T> void templateWithNonTemplatizedParameterCheapTemplate(const ExpensiveToCopyType S, T V) {
+  // CHECK-MESSAGES: [[@LINE-1]]:103: warning: the const qualified parameter 'S'
+  // CHECK-FIXES: template <typename T> void templateWithNonTemplatizedParameterCheapTemplate(const ExpensiveToCopyType& S, T V) {
+}
+
+void instantiatedWithCheapValue() {
+  templateWithNonTemplatizedParameterCheapTemplate(ExpensiveToCopyType(), 5);
+}
+
+template <typename T> void nonInstantiatedTemplateWithConstValue(const T S) {}
+template <typename T> void nonInstantiatedTemplateWithNonConstValue(T S) {}
+
+template <typename T> void instantiatedTemplateSpecialization(T NoSpecS) {}
+template <>
+void instantiatedTemplateSpecialization<ExpensiveToCopyType>(
+    ExpensiveToCopyType SpecS) {
+  // CHECK-MESSAGES: [[@LINE-1]]:25: warning: the parameter 'SpecS'
+  // When updating a template specialization, we also update the main template.
```

Comment:
I don't think that we need to do that. The specialization can specialize for the specific type, and the base template can be `const&`-ed when an instantiation contains a type that is not cheap to copy. WDYT?

Reply:
> Note that it is template specialization and not an overload.
Template specialization must have exactly the same signature.

Whoops, I must have done something weird to think this works.

Then what I would have originally written applies:
When a specialization A is defined in another TU, applying a fix to specialization B from the analyzed TU or the underlying template, will make A no longer compile. The same goes for specializations in headers that are not included, and also specializations that are:

```c++
template <typename T> void instantiatedTemplateSpecialization(T NoSpecS) {}
template <>
void instantiatedTemplateSpecialization<ExpensiveToCopyType>(
    ExpensiveToCopyType SpecS) {
  // CHECK-MESSAGES: [[@LINE-1]]:25: warning: the parameter 'SpecS'
  // When updating a template specialization, we also update the main template.
  // CHECK-FIXES: const T& NoSpecS
  // CHECK-FIXES: const ExpensiveToCopyType& SpecS
  // CHECK-FIXES: const int& SpecS               <- fails, no fix generated it seems
}
template <>
void instantiatedTemplateSpecialization<int>(
    int SpecS) {
}
```

Reply:
That is unfortunately correct.  I was aware that other TUs can be broken.

But it seems that other template instantiations are  not fixed in the same file as well.

Do you know how to skip template specialization, but not all other template functions?

Reply:
I have found a partial fix for this issue (and created fixup commit).

The changes are:
1. No fixes are proposed for explicit specializations.
2. Fixes are still proposed for the main template in case warning is issued during instantiation of the main template. I couldn't find the way to find that any instantiation is available. And in case they are defined in another translation unit, there is no way to do it.